File: 2022-06-craft-conf-languages/examples/parse_digit.go
package main

import "errors"
import "fmt"

var InvalidDigit = errors.New("not a digit")

func ParseDigit(ch byte) (int, error) {
	if ch >= '0' && ch <= '9' {
		return int(ch - '0'), nil
	}
	return 0, InvalidDigit
}

func ParseFile(contents string) error {
	_, err := ParseDigit(contents[0])
	if err != nil {
		return err
	}
	return nil
}

func UseFile(contents string) {
	err := ParseFile(contents)
	if err != nil {
		print(err)
	}
}

func main() {
	i, err := ParseDigit('2')
	fmt.Printf("%d %v\n", i, err)
	i, err = ParseDigit('a')
	fmt.Printf("%d %v\n", i, err)
}

File: 2022-06-craft-conf-languages/examples/parse_digit.rs
struct InvalidDigit;

fn parse_digit(ch: u8)
  -> Result<u8, InvalidDigit> {
    if ch >= b'0' && ch <= b'9' {
        return Ok(ch - b'0');
    }
    Err(InvalidDigit)
}

fn parse_file(contents: &[u8])
  -> Result<(), InvalidDigit> {
    parse_digit(contents[0])?;
    Ok(())
}

fn use_file(contents: &[u8]) {
    match parse_file(contents) {
        Ok(_) => {}
        Err(_) => println!("not a digit!"),
    }
}

fn main() {
    let digit = unsafe { parse_digit(b'2').unwrap_unchecked() };
    println!("{}", digit);
}
